### PR TITLE
KIP 848 ListGroups API 

### PR DESCRIPTION
**Pull request by mahajanadhitya**

### CODE REVIEW

File: src/Confluent.Kafka/Impl/NativeMethods/NativeMethods.cs [hunk 1]
```diff
@@ -742,7 +748,7 @@ long least_significant_bits
                 /* rd_kafka_ConfigEntry_t * */ IntPtr entry);
 
         [DllImport(DllName, CallingConvention = CallingConvention.Cdecl)]
-        internal static extern IntPtr rd_kafka_ConfigEntry_is_synonym (
+        internal static extern IntPtr rd_kafka_ConfigEntry_is_synonym(
```
Comment (anchitj) [inline src/Confluent.Kafka/Impl/NativeMethods/NativeMethods.cs:751]: Can we remove the changes done by style fix at places which we don't intend to touch? It's unnecessarily increasing the diff size

**Review by anchitj — CHANGES_REQUESTED** [review]

File: src/Confluent.Kafka/Impl/NativeMethods/NativeMethods.cs [hunk 2]
```diff
@@ -564,6 +564,18 @@ internal static extern IntPtr rd_kafka_AdminOptions_set_match_consumer_group_sta
...
+                        IntPtr options,
+                        ConsumerGroupType[] groupTypes,
+                        UIntPtr groupTypesCnt);
+
+        [DllImport(DllName, CallingConvention = CallingConvention.Cdecl)]
+        internal static extern IntPtr rd_kafka_AdminOptions_set_match_consumer_group_types(
+                        IntPtr options,
+                        ConsumerGroupType[] groupTypes,
+                        UIntPtr groupTypesCnt);
+
```
Comment (anchitj) [inline src/Confluent.Kafka/Impl/NativeMethods/NativeMethods.cs:578]: Why is it defined two times? This is causing CI failure? Same in other NativeMethods files

File: test/Confluent.Kafka.IntegrationTests/Tests/AdminClient_ListDescribeConsumerGroups.cs [hunk 3]
```diff
@@ -63,16 +63,68 @@ private void checkConsumerGroupDescription(
         [Theory, MemberData(nameof(KafkaParameters))]
         public void AdminClient_ListDescribeConsumerGroups(string bootstrapServers)
         {
+            var groupID = Guid.NewGuid().ToString();
+            var nonExistentGroupID = Guid.NewGuid().ToString();
```
Comment (anchitj) [inline test/Confluent.Kafka.IntegrationTests/Tests/AdminClient_ListDescribeConsumerGroups.cs:67]: Renamed to `groupId` and `nonExistentGroupId`

File: test/Confluent.Kafka.IntegrationTests/Tests/AdminClient_ListDescribeConsumerGroups.cs [hunk 4]
```diff
@@ -63,16 +63,68 @@ private void checkConsumerGroupDescription(
...
+                LogToFile("start AdminClient_ListDescribeConsumerGroups with Consumer Protocol");
+                const string clientID = "test.client";
+
+                // Create an AdminClient here - we need it throughout the test.
+                using (var adminClient = new AdminClientBuilder(new AdminClientConfig
+                {
+                    BootstrapServers = bootstrapServers
+                }).Build())
+                {
+                    var listOptionsWithTimeout = new Admin.ListConsumerGroupsOptions() { RequestTimeout = TimeSpan.FromSeconds(30) };
```
Comment (anchitj) [inline test/Confluent.Kafka.IntegrationTests/Tests/AdminClient_ListDescribeConsumerGroups.cs:79]: Not being used

File: test/Confluent.Kafka.IntegrationTests/Tests/AdminClient_ListDescribeConsumerGroups.cs [hunk 5]
```diff
@@ -63,16 +63,68 @@ private void checkConsumerGroupDescription(
...
         public void AdminClient_ListDescribeConsumerGroups(string bootstrapServers)
         {
+            var groupID = Guid.NewGuid().ToString();
+            var nonExistentGroupID = Guid.NewGuid().ToString();
             if (!TestConsumerGroupProtocol.IsClassic())
             {
-                LogToFile("KIP 848 Admin operations changes still aren't " +
-                          "available");
+                LogToFile("start AdminClient_ListDescribeConsumerGroups with Consumer Protocol");
+                const string clientID = "test.client";
```
Comment (anchitj) [inline test/Confluent.Kafka.IntegrationTests/Tests/AdminClient_ListDescribeConsumerGroups.cs:71]: clientId. Name variables using camelCase

File: test/Confluent.Kafka.IntegrationTests/Tests/AdminClient_ListDescribeConsumerGroups.cs [hunk 6]
```diff
@@ -63,16 +63,68 @@ private void checkConsumerGroupDescription(
         [Theory, MemberData(nameof(KafkaParameters))]
         public void AdminClient_ListDescribeConsumerGroups(string bootstrapServers)
         {
+            var groupID = Guid.NewGuid().ToString();
+            var nonExistentGroupID = Guid.NewGuid().ToString();
             if (!TestConsumerGroupProtocol.IsClassic())
             {
-                LogToFile("KIP 848 Admin operations changes still aren't " +
```
Comment (anchitj) [inline test/Confluent.Kafka.IntegrationTests/Tests/AdminClient_ListDescribeConsumerGroups.cs:68]: Why was this removed?

File: test/Confluent.Kafka.IntegrationTests/Tests/AdminClient_ListDescribeConsumerGroups.cs [hunk 7]
```diff
@@ -63,16 +63,68 @@ private void checkConsumerGroupDescription(
         [Theory, MemberData(nameof(KafkaParameters))]
         public void AdminClient_ListDescribeConsumerGroups(string bootstrapServers)
         {
+            var groupID = Guid.NewGuid().ToString();
```
Comment (anchitj) [inline test/Confluent.Kafka.IntegrationTests/Tests/AdminClient_ListDescribeConsumerGroups.cs:66]: Please add another similar test creating `Classic` group

File: examples/AdminClient/Program.cs [hunk 8]
```diff
@@ -583,27 +583,49 @@ static async Task ListConsumerGroupsAsync(string bootstrapServers, string[] comm
         {
             var timeout = TimeSpan.FromSeconds(30);
             var statesList = new List<ConsumerGroupState>();
-            try
+            var groupTypesList = new List<ConsumerGroupType>();
+            var isType = false;
+            var isState = false;
+            for (int i = 0; i < commandArgs.Length; i++)
```
Comment (anchitj) [inline examples/AdminClient/Program.cs:589]: Use `foreach` like `foreach (var t in commandArgs)`

File: examples/AdminClient/Program.cs [hunk 9]
```diff
@@ -583,27 +583,49 @@ static async Task ListConsumerGroupsAsync(string bootstrapServers, string[] comm
...
                     {
                         statesList.Add(Enum.Parse<ConsumerGroupState>(commandArgs[i]));
                     }
+                    else if (isType)
+                    {
+                        groupTypesList.Add(Enum.Parse<ConsumerGroupType>(commandArgs[i]));
+                    }
+                    else
+                    {
+                        Console.WriteLine("usage: .. <bootstrapServers> list-consumer-groups [-states <match_state_1> <match_state_2> ... <match_state_N>] [-types <group_type_1> .. <group_type_M>]");
```
Comment (anchitj) [inline examples/AdminClient/Program.cs:623]: Missing `<timeout_seconds>`

File: examples/AdminClient/Program.cs [hunk 10]
```diff
@@ -583,27 +583,49 @@ static async Task ListConsumerGroupsAsync(string bootstrapServers, string[] comm
...
+            var isState = false;
+            for (int i = 0; i < commandArgs.Length; i++)
             {
-                if (commandArgs.Length > 0)
+                if (commandArgs[i] == "-states")
                 {
-                    timeout = TimeSpan.FromSeconds(Int32.Parse(commandArgs[0]));
+                    if (isState)
+                    {
+                        Console.WriteLine("usage: .. <bootstrapServers> list-consumer-groups [-states <match_state_1> <match_state_2> ... <match_state_N>] [-types <group_type_1> .. <group_type_M>]");
```
Comment (anchitj) [inline examples/AdminClient/Program.cs:595]: Improve the error message, state that this was a Duplicate states argument. Same for types

**Review by anchitj — CHANGES_REQUESTED** [review]
Please also make the CI pass, and confirm that you've manually tested the code locally.

File: src/Confluent.Kafka/Impl/NativeMethods/NativeMethods.cs (hunk 2 above)
Comment (mahajanadhitya) [reply src/Confluent.Kafka/Impl/NativeMethods/NativeMethods.cs:578]: Actually these were due to rebasing of the "squash & merge" commits and these did not raise any merge conflict as well. So after local build/test and format, the last rebase messed these up. Would take care for next time !

File: test/Confluent.Kafka.IntegrationTests/Tests/AdminClient_ListDescribeConsumerGroups.cs (hunk 3 above)
Comment (mahajanadhitya) [reply test/Confluent.Kafka.IntegrationTests/Tests/AdminClient_ListDescribeConsumerGroups.cs:67]: ok

File: test/Confluent.Kafka.IntegrationTests/Tests/AdminClient_ListDescribeConsumerGroups.cs (hunk 5 above)
Comment (mahajanadhitya) [reply test/Confluent.Kafka.IntegrationTests/Tests/AdminClient_ListDescribeConsumerGroups.cs:71]: since it was id only then kept d capital as it is a short form. But will reflect this in the change.

File: test/Confluent.Kafka.IntegrationTests/Tests/AdminClient_ListDescribeConsumerGroups.cs (hunk 7 above)
Comment (mahajanadhitya) [reply test/Confluent.Kafka.IntegrationTests/Tests/AdminClient_ListDescribeConsumerGroups.cs:66]: Actually the way 848 integration testing work is that we run 2 instances of Semaphore CI where one version runs for all the test having no broker min version, so assume only classic group protocol is used. 
Now currently semaphore will only run some parts of test that needs verifying which means with broker version > 3.8.0.0 and environment variable for the consumer group protocol chosen to be consumer type.

File: examples/AdminClient/Program.cs (hunk 8 above)
Comment (mahajanadhitya) [reply examples/AdminClient/Program.cs:589]: ok

File: examples/AdminClient/Program.cs (hunk 9 above)
Comment (mahajanadhitya) [reply examples/AdminClient/Program.cs:623]: None of the usage in codebase mentions timeout seconds.

Comment (mahajanadhitya) [reply examples/AdminClient/Program.cs:623]: For the examples convention I am saying

File: test/Confluent.Kafka.IntegrationTests/Tests/AdminClient_ListDescribeConsumerGroups.cs (hunk 6 above)
Comment (mahajanadhitya) [reply test/Confluent.Kafka.IntegrationTests/Tests/AdminClient_ListDescribeConsumerGroups.cs:68]: Because these will be operational once the PR goes in and even in preview mode if unsafe api kafka server setting is set, then these are operational.

File: examples/AdminClient/Program.cs (hunk 10 above)
Comment (mahajanadhitya) [reply examples/AdminClient/Program.cs:595]: No the duplicate state is Invalid args but still the example will run and yield a error like Invalid args.
This is basically error with exit code so user has put in wrong input format for the example.
like -state_cnt=2 state_1 -group_type_cnt=..
The above will error because only 1 state was given.

Comment (mahajanadhitya) [reply examples/AdminClient/Program.cs:595]: Please suggest something if you feel it is still not sufficient.

File: examples/AdminClient/Program.cs (hunk 9 above)
Comment (anchitj) [reply examples/AdminClient/Program.cs:623]: See the remove line 602 it had `< timeout_seconds>`

**Comment by mahajanadhitya** [comment]
**confluent-kafka-dotnet KIP 848 ListGroups API**
Branch Name : `feature/ListGroupsAPI`
Example : `confluent-kafka-dotnet/examples/AdminClient/Program.cs`

Running the example : `dotnet run <bootstrapServers> list-consumer-groups [-states <match_state_1> <match_state_2> ... <match_state_N>] [-types <group_type_1> .. <group_type_M>]`
Types can only be Consumer(1), Classic(2). Any other type int would give error as Unknown(0) is not accepted and others do not exist
Types only populates the options for the request which MIGHT NOT BE USED by broker, depending on the version used.

**Make sure to install the Librdkafka `feature/ListGroupsAPI` and making the `dotnet build` with the local librdkafka**

**Any warnings via configure, make or make install are not subject to my changes.**

Formatting done via, 
`$ dotnet format`

Spawn the trivup cluster from `librdkafka` via:
`$ ./interactive_broker_version.py 'trunk@f6c9feea76d01a46319b0ca602d70aa855057b07' --kraft --conf '{"conf":["unstable.api.versions.enable=true","group.coordinator.new.enable=true","group.protocol=classic,consumer"]}'`

Integration Testing done via :
**Consumer**
`$ export TEST_CONSUMER_GROUP_PROTOCOL=consumer` 
`$ dotnet test --filter "FullyQualifiedName=Confluent.Kafka.IntegrationTests.Tests.AdminClient_ListDescribeConsumerGroups"`
**Classic**
`$ export TEST_CONSUMER_GROUP_PROTOCOL=classic` 
`$ dotnet test --filter "FullyQualifiedName=Confluent.Kafka.IntegrationTests.Tests.AdminClient_ListDescribeConsumerGroups"`

**Comment by emasab** [comment]
Closed in favor of #2323 because auto format will be applied in a different PR
